### PR TITLE
 Add pod-based metric_labels to destinations response (#429)

### PR DESCRIPTION
**Pull request by pcalcado**
This builds on #645, the only commit it adds is 9c2c4ea so you might want to look at it in isolation.

It adds the `k8s_namespace`, `k8s_service`,  and `k8s_pod` `metric_labekls` [as described in #429 ](https://github.com/runconduit/conduit/issues/429#issuecomment-376224043). A run with the test script looks like this:

```
 $ bin/go-run  controller/script/destination-client/main.go -path stat-test-t2-svc.default.svc.cluster.local:90
command-line-arguments
INFO[0000] Add:
INFO[0000] metric_labels: map[k8s_service:stat-test-t2-svc.default.svc.cluster.local:90 k8s_namespace:default]
INFO[0000] - 172.17.0.28:90 - map[k8s_pod:stat-test-t2-6879cfb7b5-qcjtv]
INFO[0000] - 172.17.0.45:90 - map[k8s_pod:stat-test-t2-6879cfb7b5-8t5l8]
INFO[0000]
```

### CODE REVIEW

**Comment by adleong** [comment]
I don't think we want to send the `k8s_pod` label as this will cause the proxy to expose a `dst_k8s_pod` label which will be of large cardinality.  Instead, I think we should be setting `k8s_job` or `k8s_replicaset` or `k8s_deployment`... etc.  These can be found by using `pod.GetOwnerReferences()`.  See `GetDeploymentForPod` in `replicasets.go` for an example.

File: controller/destination/destination_resolver.go [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+		addrs = append(addrs, &pb.WeightedAddr{
+			Addr:         &endpoints[i],
+			Weight:       1,
+			MetricLabels: metricLabelsForPod,
+		})
+	}
+
+	globalMetricLabels := map[string]string{
+		"k8s_service":   serviceName,
+		"k8s_namespace": namespace,
```
Comment (siggy) [inline controller/destination/destination_resolver.go]: we've renamed these to `namespace`, have a look at `proxy-metrics.md`.
re: `k8s_pod` above, we just added `pod_name` to `proxy-metrics.md`, but i'm realizing it should probably just be `pod` to be consistent with the other k8s objects (`deployment`, `namespace`, etc)
also related: i'm realizing we left off `service` and `dst_service` from `proxy-metrics.md`.

File: controller/script/destination-client/main.go [hunk 2]
```diff
@@ -51,8 +51,9 @@ func main() {
 		switch updateType := update.Update.(type) {
 		case *pb.Update_Add:
 			log.Println("Add:")
+			log.Printf("metric_labels: %v", updateType.Add.MetricLabels)
 			for _, addr := range updateType.Add.Addrs {
-				log.Printf("- %s:%d", util.IPToString(addr.Addr.GetIp()), addr.Addr.Port)
+				log.Printf("- %s:%d - %v", util.IPToString(addr.Addr.GetIp()), addr.Addr.Port, addr.MetricLabels)
```
Comment (siggy) [inline controller/script/destination-client/main.go]: note totally related to this change, but would you mind mentioning this tool in `BUILD.md`? maybe in the `Go` section, just something short like:

```bash
# forward localhost:8089 to the destination service
$ kubectl -n conduit port-forward $(kubectl --namespace=conduit get po --selector=conduit.io/control-plane-component=controller -o jsonpath='{.items[*].metadata.name}') 8089:8089

# test connecting to the destination service
$ bin/go-run  controller/script/destination-client/main.go -path stat-test-t2-svc.default.svc.cluster.local:90
```

also, at the top of this file, mind changing:
```go
addr := flag.String("addr", ":8089", "address of proxy api")
```
to:
```go
addr := flag.String("addr", ":8089", "address of destination service")
```

**Comment by adleong** [comment]
Oh, interesting!  @olix0r do you think `dst_k8s_pod` is a good idea?  I thought we had concluded that that would be too much.

File: controller/destination/destination_resolver.go (hunk 1 above)
Comment (siggy) [reply controller/destination/destination_resolver.go]: taking this discussion over to #655

File: controller/script/destination-client/main.go (hunk 2 above)
Comment (pcalcado) [reply controller/script/destination-client/main.go]: @siggy wdyt https://github.com/runconduit/conduit/pull/654/commits/93a56b94f79e7f7342aee987119dc4d59b2362ed

**Comment by hawkw** [comment]
This will, of course, necessitate some additional changes on the Rust side, but for now, we should be able to fix the build by making the following change:
```patch


From 28508e6f278a0b8f4f5956149b8528b328f88976 Mon Sep 17 00:00:00 2001
From: Eliza Weisman <eliza@buoyant.io>
Date: Fri, 30 Mar 2018 10:57:13 -0700
Subject: [PATCH] Fix mock controller in proxy tests

---
 proxy/tests/support/controller.rs | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/proxy/tests/support/controller.rs b/proxy/tests/support/controller.rs
index 220b474..376a79e 100644
--- a/proxy/tests/support/controller.rs
+++ b/proxy/tests/support/controller.rs
@@ -281,8 +281,10 @@ pub fn destination_update(addr: SocketAddr) -> pb::destination::Update {
                             port: u32::from(addr.port()),
                         }),
                         weight: 0,
+                        ..Default::default()
                     },
                 ],
+                ..Default::default()
             },
         )),
     }
-- 
2.14.3 (Apple Git-98)
``

**Pull request merged**